File: cpp-sudoku/src/generator/Generator.cpp
#include "Generator.hpp"
#include "../util/SudokuUtils.hpp"
#include <algorithm>
#include <random>
#include <cmath>

namespace just::demo::generator {

Generator::Generator(int size, int complexityLowerLimit) 
    : size_(size), complexityLowerLimit_(complexityLowerLimit), blockSize_(static_cast<int>(std::sqrt(size))) {
    for (int i = 1; i <= size; ++i) {
        values_.insert(i);
    }
}

std::vector<std::vector<int>> Generator::generate() {
    std::vector<std::vector<int>> initialValues(size_, std::vector<int>(size_, 0));
    return generate(initialValues);
}

std::vector<std::vector<int>> Generator::generate(const std::vector<std::vector<int>>& initialValues) {
    if (just::demo::util::SudokuUtils::countOpen(initialValues) > complexityLowerLimit_) {
        throw just::demo::exception::ComplexityLimitException();
    }

    try {
        just::demo::solver::Solver(initialValues).solve();
        return initialValues;
    } catch (const just::demo::exception::MultipleSolutionsException&) {
        std::vector<Cell> open;
        std::vector<Cell> pending;
        
        for (int row = 0; row < size_; ++row) {
            for (int col = 0; col < size_; ++col) {
                int value = initialValues[row][col];
                int block = blockSize_ * (row / blockSize_) + col / blockSize_;
                Cell cell(row, col, block, value);
                if (value != 0) {
                    open.push_back(cell);
                } else {
                    pending.push_back(cell);
                }
            }
        }

        // Shuffle pending cells
        std::random_device rd;
        std::mt19937 g(rd());
        std::shuffle(pending.begin(), pending.end(), g);
        
        for (const auto& cell : pending) {
            auto nextGuess = just::demo::util::SudokuUtils::copy(initialValues);
            std::vector<int> availableValues(values_.begin(), values_.end());
            
            // Remove values from related cells
            for (const auto& openCell : open) {
                if (cell.isRelated(openCell)) {
                    availableValues.erase(
                        std::remove(availableValues.begin(), availableValues.end(), openCell.value),
                        availableValues.end()
                    );
                }
            }
            
            // Shuffle available values
            std::shuffle(availableValues.begin(), availableValues.end(), g);
            
            for (int value : availableValues) {
                nextGuess[cell.row][cell.col] = value;
                try {
                    return generate(nextGuess);
                } catch (const just::demo::exception::NoSolutionException&) {
                    // Our guess did not work, let's try another one
                }
            }
        }
        throw just::demo::exception::NoSolutionException();
    }
}

} // namespace just::demo::generator

File: src/main/java/just/demo/generator/Generator.java
package just.demo.generator;

import static java.util.Collections.shuffle;
import static java.util.stream.Collectors.toSet;
import static java.util.stream.IntStream.rangeClosed;

import static just.demo.util.SudokuUtils.copy;
import static just.demo.util.SudokuUtils.countOpen;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import just.demo.exception.ComplexityLimitException;
import just.demo.exception.MultipleSolutionsException;
import just.demo.exception.NoSolutionException;
import just.demo.solver.Solver;
import lombok.Value;

public class Generator {

  private final int size;
  private final int complexityLowerLimit;
  private final int blockSize;
  private final Set<Integer> values;

  public Generator(int size, int complexityLowerLimit) {
    this.size = size;
    this.complexityLowerLimit = complexityLowerLimit;
    this.blockSize = (int) Math.sqrt(size);
    this.values = rangeClosed(1, size).boxed().collect(toSet());
  }

  public int[][] generate() {
    return generate(new int[size][size]);
  }

  private int[][] generate(int[][] initialValues) {
    if (countOpen(initialValues) > complexityLowerLimit) {
      throw new ComplexityLimitException();
    }

    try {
      new Solver(initialValues).solve();
      return initialValues;
    } catch (MultipleSolutionsException e) {
      List<Cell> open = new ArrayList<>();
      List<Cell> pending = new ArrayList<>();
      for (int row = 0; row < size; row++) {
        for (int col = 0; col < size; col++) {
          int value = initialValues[row][col];
          int block = blockSize * (row / blockSize) + col / blockSize;
          Cell cell = new Cell(row, col, block, value);
          if (value != 0) {
            open.add(cell);
          } else {
            pending.add(cell);
          }
        }
      }

      shuffle(pending);
      for (Cell cell : pending) {
        int[][] nextGuess = copy(initialValues);
        List<Integer> values = new ArrayList<>(this.values);
        open.stream().filter(cell::isRelated).map(Cell::getValue).forEach(values::remove);
        shuffle(values);
        for (Integer value : values) {
          nextGuess[cell.getRow()][cell.getCol()] = value;
          try {
            return generate(nextGuess);
          } catch (NoSolutionException e2) {
            // Our guess did not work, let's try another one
          }
        }
      }
      throw new NoSolutionException();
    }
  }

  @Value
  private static class Cell {

    int row;
    int col;
    int block;
    int value;

    public boolean isRelated(Cell cell) {
      return row == cell.row || col == cell.col || block == cell.block;
    }
  }
}
